libgalois/include/katana/PropertyGraph.h

```diff
@@ -767,6 +819,620 @@ CreateSymmetricGraph(PropertyGraph* pg);
+    // EdgeLabelIndex edge_label_index = EdgeLabelIndex::Make(pg);
+    // OutTopoBase out_topo = OutTopoBase::MakeFromDefaultTopology(pg, &edge_label_index);
+    // EdgeLabelAwareTopology in_topo = EdgeLabelAwareTopology::MakeFromTransposeTopology(pg, &edge_label_index);
+    // return EdgeLabelAwareBiDirTopology{std::move(edge_label_index), std::move(out_topo), std::move(in_topo)};
+
+    auto edge_label_index =
+        std::make_unique<EdgeLabelIndex>(EdgeLabelIndex::Make(pg));
+    return EdgeLabelAwareBiDirTopology{pg, std::move(edge_label_index)};
+  }
+
+  EdgeLabelAwareBiDirTopology() = default;
+  EdgeLabelAwareBiDirTopology(EdgeLabelAwareBiDirTopology&&) = default;
+  EdgeLabelAwareBiDirTopology& operator=(EdgeLabelAwareBiDirTopology&&) =
+      default;
+
+  EdgeLabelAwareBiDirTopology(const EdgeLabelAwareBiDirTopology&) = delete;
+  EdgeLabelAwareBiDirTopology& operator=(const EdgeLabelAwareBiDirTopology&) =
+      delete;
+
+  bool has_edge_label(const EdgeLabelTy& label) const noexcept {
+    return edge_label_index_->has_edge_label(label);
+  }
+
+  auto distinct_edge_labels() const noexcept {
+    return edge_label_index_->distinct_edge_labels();
+  }
+
+  edge_iterator in_edge_begin(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.edge_begin(N, label);
+  }
+
+  edge_iterator in_edge_end(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.edge_end(N, label);
+  }
+
+  edge_iterator in_edge_begin(Node N) const noexcept {
+    return in_topo_.edge_begin(N);
+  }
+
+  edge_iterator in_edge_end(Node N) const noexcept {
+    return in_topo_.edge_end(N);
+  }
+
+  edges_range in_edges(Node N) const noexcept { return in_topo_.edges(N); }
+
+  edges_range in_edges(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.edges(N, label);
+  }
+
+  Node in_edge_dest(Edge edge_id) const noexcept {
+    return in_topo_.edge_dest(edge_id);
+  }
+
+  Node in_edge_dest(const edge_iterator ei) const noexcept {
+    return in_topo_.edge_dest(ei);
+  }
+  size_t in_degree(Node N) const noexcept { return in_topo_.degree(N); }
+
+  size_t in_degree(Node N, const EdgeLabelTy& label) const noexcept {
+    return in_topo_.degree(N, label);
+  }
+
+  Edge original_edge_id_using_in_edge(
+      const edge_iterator& in_edge_iter) const noexcept {
+    return in_topo_.original_edge_id(in_edge_iter);
+  }
+
+  Edge original_edge_id_using_in_edge(Edge in_edge) const noexcept {
+    return in_topo_.original_edge_id(in_edge);
+  }
+
+  auto edge_label_using_in_edge(Edge eid) const noexcept {
+    return in_topo_.edge_label(eid);
+  }
+
+  /**
+   * Returns an edge iterator to an edge with some node and key by
+   * searching for the key via the node's outgoing or incoming edges.
+   * If not found, returns nothing.
+   */
+  std::optional<edges_range> FindAllEdges(Node src, Node dst) const {
+    if (OutTopoBase::degree(src) == 0 || in_topo_.degree(dst) == 0) {
+      return std::nullopt;
```

Comment:
Why not return an empty range?

Reply:
Let me think about that. I lifted these functions from LC_CSR_CSC_Labeled_Graph, which was being used by libquery's AttributedGraph.

Reply:
we have empty range now.